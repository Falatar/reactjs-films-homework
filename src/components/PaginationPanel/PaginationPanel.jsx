import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import style from './PaginationPanel.scss';
import first from '../../static/first.svg';
import previous from '../../static/previous.svg';
import next from '../../static/next.svg';
import last from '../../static/last.svg';

function PaginationPanel({
  actualPage,
  totalFilms,
  left,
  right,
  finallyLeft,
  finallyRight,
  activeMode,
  activeGenre,
  activeView,
  searchStr,
}) {
  const filmsOnPage = 12;
  const numberOfPages = parseInt((totalFilms / filmsOnPage).toFixed(0), 10);
  return (
    <div className={style.pagination__panel}>
      <button
        type="button"
        className={style.pagination__left__to__end}
        onClick={finallyLeft}
        disabled={actualPage === 1}
      >
        <Link
          to={!searchStr.length
            ? `/1/${activeView}/${activeMode}${activeGenre !== 0 ? `/${activeGenre}` : ''}`
            : `/search/${searchStr}/1/${activeView}`}
          className={style.link}
        >
          <svg className={style.svg}>
            <use xlinkHref={first} />
          </svg>
        </Link>
      </button>
      <button
        type="button"
        className={style.pagination__left}
        onClick={() => left(numberOfPages)}
        disabled={actualPage === 1}
      >
        <Link
          to={!searchStr.length
            ? `/${actualPage - 1 < 1 ? 1 : actualPage - 1}/${activeView}/${activeMode}${activeGenre !== 0 ? `/${activeGenre}` : ''}`
            : `/search/${searchStr}/${actualPage - 1 < 1 ? 1 : actualPage - 1}/${activeView}`}
          className={style.link}
        >
          <svg className={style.svg}>
            <use xlinkHref={previous} />
          </svg>
        </Link>
      </button>
      <input
        type="text"
        value={actualPage}
        className={style.pagination__to__certain__page}
        readOnly
      />
      <button
        type="button"
        className={style.pagination__right}
        onClick={() => right(numberOfPages)}
        disabled={actualPage === numberOfPages}
      >
        <Link
          to={!searchStr.length
            ? `/${actualPage + 1 > numberOfPages ? numberOfPages : actualPage + 1}/${activeView}/${activeMode}${activeGenre !== 0 ? `/${activeGenre}` : ''}`
            : `/search/${searchStr}/${actualPage + 1 > numberOfPages ? numberOfPages : actualPage + 1}/${activeView}`}
          className={style.link}
        >
          <svg className={style.svg}>
            <use xlinkHref={next} />
          </svg>
        </Link>
      </button>
      <button
        type="button"
        className={style.pagination__right__to__end}
        onClick={() => finallyRight(numberOfPages)}
        disabled={actualPage === numberOfPages}
      >
        <Link
          to={!searchStr.length
            ? `/${numberOfPages}/${activeView}/${activeMode}${activeGenre !== 0 ? `/${activeGenre}` : ''}`
            : `/search/${searchStr}/${numberOfPages}/${activeView}`}
          className={style.link}
        >
          <svg className={style.svg}>
            <use xlinkHref={last} />
          </svg>
        </Link>
      </button>
    </div>
  );
}

PaginationPanel.defaultProps = {
  totalFilms: 1,
  actualPage: 1,
  left: () => 'can\'t find left function',
  right: () => 'can\'t find right function',
  finallyLeft: () => 'can\'t find finallyLeft function',
  finallyRight: () => 'can\'t find finallyRight function',
  activeMode: '',
  activeGenre: '',
  activeView: '',
  searchStr: '',
};

PaginationPanel.propTypes = {
  totalFilms: PropTypes.number,
  actualPage: PropTypes.number,
  left: PropTypes.func,
  right: PropTypes.func,
  finallyLeft: PropTypes.func,
  finallyRight: PropTypes.func,
  activeMode: PropTypes.string,
  activeGenre: PropTypes.string,
  activeView: PropTypes.string,
  searchStr: PropTypes.string,
};

export default PaginationPanel;
