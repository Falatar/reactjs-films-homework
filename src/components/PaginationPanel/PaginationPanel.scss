.pagination__panel {
    width: 100%;
    height: 7vh;
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    background-color: lightgrey;

    .pagination__left, 
    .pagination__left__to__end, 
    .pagination__right, 
    .pagination__right__to__end {
        height: 1.6em;
        width: 3.2em;
        border-radius: 15px;
        outline: none;
        border: 0;
        padding: 0;
        //padding-bottom: 4%;
        margin: 0 0.5%;
        background-color: deepskyblue;
        color: white;
        font-size: 1.5em;

        .icon__left__to__end,
        .icon__left,
        .icon__right,
        .icon__right__to__end {
            position: relative;
            display: block;
            width: 100px;
            height: 100px;
            stroke-width: 0;
            stroke: white;
            fill: white;
        }
    }

    .pagination__left:active, 
    .pagination__left__to__end:active, 
    .pagination__right:active, 
    .pagination__right__to__end:active {
        background-image: -webkit-linear-gradient(bottom, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.4));
        background-image: -moz-linear-gradient(bottom, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.4));
        background-image: -o-linear-gradient(bottom, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.4));
        background-image: -ms-linear-gradient(bottom, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.4));
        background-image: linear-gradient(bottom, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.4));
    }

    .pagination__left:disabled, 
    .pagination__left__to__end:disabled, 
    .pagination__right:disabled, 
    .pagination__right__to__end:disabled {
        background-image: -webkit-linear-gradient(bottom, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.4));
        background-image: -moz-linear-gradient(bottom, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.4));
        background-image: -o-linear-gradient(bottom, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.4));
        background-image: -ms-linear-gradient(bottom, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.4));
        background-image: linear-gradient(bottom, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.4));
    }

    .pagination__to__certain__page {
        height: 1.2em;
        width: 5em;
        font-size: 1.5em;
        border-radius: 15px;
        text-align: center;
        color: deepskyblue;
        border: 3px solid;
        border-color: deepskyblue;
        background-color: white;
        outline: none;
    }
}