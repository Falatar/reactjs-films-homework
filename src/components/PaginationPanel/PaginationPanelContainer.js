import { connect } from 'react-redux';
import {
  moveLeft,
  moveRight,
  moveLeftToEnd,
  moveRightToEnd,
} from '../../modules/filmListModule/filmListModuleActions';
import { getNumberOfFilms, getCurrentPage } from '../../modules/filmListModule/filmListModuleSelector';
import PaginationPanel from './PaginationPanel';

const mapStateToProps = (store) => ({
  totalFilms: getNumberOfFilms(store),
  actualPage: getCurrentPage(store),
});

const mapDispatchToProps = (dispatch) => ({
  left: (value) => dispatch(moveLeft(value)),
  right: (value) => dispatch(moveRight(value)),
  finallyLeft: () => dispatch(moveLeftToEnd()),
  finallyRight: (value) => dispatch(moveRightToEnd(value)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(PaginationPanel);
