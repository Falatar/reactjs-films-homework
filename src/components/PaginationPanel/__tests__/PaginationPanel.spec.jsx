import React from 'react';
import renderer from 'react-test-renderer';
import PaginationPanel from '../PaginationPanel';

jest.mock('react-router-dom', () => {
  const originalModule = jest.requireActual('react-router-dom');

  return {
    __esModule: true,
    ...originalModule,
    Link: 'Link',
    useParams: jest.fn(),
    useHistory: jest.fn(),
  };
});

describe('PaginationPanel', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<PaginationPanel totalFilms={1} actualPage={1} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('button 2 works correctly', () => {
    const left = jest.fn();
    const component = renderer.create(
      <PaginationPanel totalFilms={1} actualPage={1} left={left} />,
    );
    const buttons = component.root.findAllByType('button');
    expect(buttons.length).toBe(4);
    buttons[1].props.onClick();
    expect(left).toHaveBeenCalled();
  });

  it('button 3 works correctly', () => {
    const right = jest.fn();
    const component = renderer.create(
      <PaginationPanel totalFilms={1} actualPage={1} right={right} />,
    );
    const buttons = component.root.findAllByType('button');
    expect(buttons.length).toBe(4);
    buttons[2].props.onClick();
    expect(right).toHaveBeenCalled();
  });

  it('button 4 works correctly', () => {
    const finallyRight = jest.fn();
    const component = renderer.create(
      <PaginationPanel totalFilms={1} actualPage={1} finallyRight={finallyRight} />,
    );
    const buttons = component.root.findAllByType('button');
    expect(buttons.length).toBe(4);
    buttons[3].props.onClick();
    expect(finallyRight).toHaveBeenCalled();
  });

  it('default functions works correctly', () => {
    const component = renderer.create(
      <PaginationPanel
        totalFilms={1}
        actualPage={1}
        left={undefined}
        right={undefined}
        finallyLeft={undefined}
        finallyRight={undefined}
      />,
    );
    const result1 = component.root.props.left();
    const result2 = component.root.props.right();
    const result3 = component.root.props.finallyLeft();
    const result4 = component.root.props.finallyRight();
    expect(result1).toEqual('can\'t find left function');
    expect(result2).toEqual('can\'t find right function');
    expect(result3).toEqual('can\'t find finallyLeft function');
    expect(result4).toEqual('can\'t find finallyRight function');
  });
});
