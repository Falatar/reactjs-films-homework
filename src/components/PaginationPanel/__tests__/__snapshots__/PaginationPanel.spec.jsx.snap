// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationPanel renders correctly 1`] = `
<div
  className="pagination__panel"
>
  <button
    className="pagination__left__to__end"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    <Link
      className="link"
      to="/1///"
    >
      <svg
        className="svg"
      >
        <use
          xlinkHref="mocked-image"
        />
      </svg>
    </Link>
  </button>
  <button
    className="pagination__left"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    <Link
      className="link"
      to="/1///"
    >
      <svg
        className="svg"
      >
        <use
          xlinkHref="mocked-image"
        />
      </svg>
    </Link>
  </button>
  <input
    className="pagination__to__certain__page"
    readOnly={true}
    type="text"
    value={1}
  />
  <button
    className="pagination__right"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <Link
      className="link"
      to="/0///"
    >
      <svg
        className="svg"
      >
        <use
          xlinkHref="mocked-image"
        />
      </svg>
    </Link>
  </button>
  <button
    className="pagination__right__to__end"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <Link
      className="link"
      to="/0///"
    >
      <svg
        className="svg"
      >
        <use
          xlinkHref="mocked-image"
        />
      </svg>
    </Link>
  </button>
</div>
`;
